chain/network/src/store/schema/mod.rs

```diff
@@ -0,0 +1,279 @@
+    status: KnownPeerStatus,
+    /// UNIX timestamps in nanos.
+    first_seen: u64,
+    last_seen: u64,
+}
+
+impl BorshRepr for KnownPeerStateRepr {
+    type T = primitives::KnownPeerState;
+    fn to_repr(s: &primitives::KnownPeerState) -> Self {
+        Self {
+            peer_info: s.peer_info.clone(),
+            status: s.status.clone().into(),
+            first_seen: s.first_seen,
+            last_seen: s.last_seen,
+        }
+    }
+
+    fn from_repr(s: Self) -> Result<primitives::KnownPeerState, Error> {
+        Ok(primitives::KnownPeerState {
+            peer_info: s.peer_info,
+            status: s.status.into(),
+            first_seen: s.first_seen,
+            last_seen: s.last_seen,
+        })
+    }
+}
+
+#[derive(BorshSerialize, BorshDeserialize)]
+pub struct EdgeRepr {
+    key: (PeerId, PeerId),
+    nonce: u64,
+    signature0: Signature,
+    signature1: Signature,
+    removal_info: Option<(bool, Signature)>,
+}
+
+impl BorshRepr for EdgeRepr {
+    type T = primitives::Edge;
+
+    fn to_repr(e: &Self::T) -> Self {
+        Self {
+            key: e.key().clone(),
+            nonce: e.nonce(),
+            signature0: e.signature0().clone(),
+            signature1: e.signature1().clone(),
+            removal_info: e.removal_info().cloned(),
+        }
+    }
+    fn from_repr(e: Self) -> Result<Self::T, Error> {
+        Ok(primitives::Edge::new(e.key.0, e.key.1, e.nonce, e.signature0, e.signature1)
+            .with_removal_info(e.removal_info))
+    }
+}
+
+/////////////////////////////////////////////
+// Columns
+
+pub struct AccountAnnouncements;
+impl Column for AccountAnnouncements {
+    const COL: DBCol = DBCol::AccountAnnouncements;
+    type Key = AccountIdFormat;
+    type Value = Borsh<AnnounceAccount>;
+}
+
+pub struct Peers;
+impl Column for Peers {
+    const COL: DBCol = DBCol::Peers;
+    type Key = Borsh<PeerId>;
+    type Value = KnownPeerStateRepr;
+}
+
+pub struct PeerComponent;
+impl Column for PeerComponent {
+    const COL: DBCol = DBCol::PeerComponent;
+    type Key = Borsh<PeerId>;
+    type Value = Borsh<u64>;
+}
+
+pub struct ComponentEdges;
+impl Column for ComponentEdges {
+    const COL: DBCol = DBCol::ComponentEdges;
+    type Key = U64LE;
+    type Value = Vec<EdgeRepr>;
+}
+
+pub struct LastComponentNonce;
+impl Column for LastComponentNonce {
+    const COL: DBCol = DBCol::LastComponentNonce;
+    type Key = Borsh<()>;
+    type Value = Borsh<u64>;
+}
+
+////////////////////////////////////////////////////
+// Storage
+
+// Parsing error.
+#[derive(Error, Debug)]
+pub enum Error {
+    #[error(transparent)]
+    Utf8(std::str::Utf8Error),
+    #[error(transparent)]
+    IO(std::io::Error),
+    #[error(transparent)]
+    Slice(std::array::TryFromSliceError),
+    #[error(transparent)]
+    AccountId(ParseAccountError),
+    #[error(transparent)]
+    Time(time::error::ComponentRange),
+}
+
+pub trait Format {
+    type T;
+    fn to_vec(a: &Self::T) -> Vec<u8>;
```

Comment:
```suggestion
    fn to_vec(a: &Self::T, buf: &mut Vec<u8>);
```

Its idiomatic rust *not* to enforce potentially avoidable allocations in the API. It's still fine to provide convenience utility fns thogh:

```rust
trait Format {
   type T;
   fn encode(value: &T, buf: &mut Vec<u8>);
   fn decode(slice: &[u8]) -> Result<T>,
   fn to_vec(value: &T) -> Vec<u8> { 
     let mut res = Vec::new();
     Self::encode(value, &mut res);
     res
   }
}
```

Reply:
Looks like a premature optimization to me. Do you genuinely expect anyone to use the encode directly rather than through to_vec? encode has a weaker type than to_vec because it accepts extra input data, it is not clear if it is supposed only to append new data, and some implementation of Format may override the default implementation of to_vec and consistency between encode and to_vec is not enforced in any way.

Reply:
To me, this is more about idioms, rather than performance. APIs which can trivially avoid forcing an allocation, but do, look odd in Rust.

Though in this case I would argue performance does matter as well. This is our db access layer, it can be bottleneck, and it is the foundation of the stuff. Would be a shame if we extend the Format pattern with `Vec` to the whole database, only to realize later that we *do* want to refactor all Formats to use a pre-allocated buffer. 

(by the same token, we should change the API of `StoreUpdate` to take vecs rather than slices -- copy there is not helpful)

Overall, I do feel we do to much gratuitous copying/allocation in nearcore, and at least once I observed that to be a bottleneck for a particular workflow https://github.com/near/nearcore/pull/7012/commits/4cfcb71870f0ce99a08ce40549e65f3d4e9f575a#diff-500ea46e5bd2bd127356f199ea3e4c267ea47a2e571c3390b5204378f15ce844R287

---

Very much agree with the point of method being overridable -- I actually consider methods with default impls to be an anti-pattern: they mix together "interface for the implementer" and "interface for the user". Here, the best soln would probably be a helper top-level function in the module which actually uses encoding.

Reply:
> APIs which can trivially avoid forcing an allocation, but do, look odd in Rust.

Let me repeat the question: do you expect anyone to **use** this non-forcing function? Because you say it is OK to have a forcing helper function. I already have it. We can expose the additional unused API surface which will require an extra section of documentation to precisely specify the expected semantics (to compensate for the weaker typing) just for the sake of an idiom, but does it really make sense?

Reply:
done

Reply:
I am thinking not in terms of users of this API, but in terms of implementors. Each implementation would like  to specify the minimum amount of stuff -- just the bytes that the format emits. 

Than there's convenience API on top, which takes the minimal API (emitting some bytes) and makes it convenient for the common case (arranging the bytes to be shoved into a freshly allocated vector). In a sense, this gets rid of the duplication of `Vec::new` in every impl. I don't think there's any unused code here, just some finer-grained factoring .

Like the solution with `Write`!